h/views/account_signup.py

```diff
@@ -219,7 +246,38 @@ def settings(self) -> SocialLoginSignupViewsSettings:
     @view_config(request_method="GET", route_name="signup.google")
     @view_config(request_method="GET", route_name="signup.facebook")
     def get(self):
-        return {"js_config": self.js_config(self.decode_idinfo().sub)}
+        idinfo = self.decode_idinfo()
+
+        # If we get to this view there normally won't be an existing account
+        # already connected to the user's third-party identity: if such an
+        # account existed the user would have been logged into it rather than
+        # being redirected to this signup page.
+        #
+        # But concurrent signup requests can mean that we get here and find
+        # that a connected account already exists in the DB.
+        #
+        # We *could* log the user in to the existing account at this point. But
+        # I think this case is going to be pretty rare so for simplicity's sake
+        # I'm just going to raise an error.
+        if self.user_service.fetch_by_identity(
+            self.settings.provider, provider_unique_id=idinfo.sub
+        ):
+            raise IdentityConflictError
+
+        # Error if there's already a Hypothesis account with the same email
+        # address as the user's third-party identity.
+        #
+        # This can happen if the user manually creates an account with the same
+        # email address as they use with their third-party identity, and then
+        # later tries to log in or sign up with that third-party identity.
+        #
+        # It can also happen if there are concurrent signup requests.
+        if User.get_by_email(
+            self.request.db, idinfo.email, self.request.default_authority
+        ):
+            raise EmailConflictError
```

Comment:
This is the view for the `GET` request to the interstitial social login signup page `/signup/<PROVIDER>`. The reason for the additional conflict checks here is just that if we know the signup is definitely going to conflict it seems nicer to error immediately, rather than rendering the form and erroring when the user fills it in and submits it.